### PR TITLE
docker

### PR DESCRIPTION
**Pull request by mistafunk**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -99,51 +103,61 @@ _Note: on Windows, replace `bin` with `Scripts` in the following commands. Some
...
 1. Install PyPRT in development mode as described [above](#iterative-python-development).
 1. Run `python setup.py build_doc`, this will output the html files in the `build/sphinx` directory.
 1. Leave development mode also as described [above](#iterative-python-development).
 
+### Docker
+
+1. Open a shell in the PyPRT git root
+1. Create the desired image for the build toolchain (adapt `centos7/py36` with your desired OS/Python combination): `docker build --rm -f envs/centos7/py36/Dockerfile -t pyprt:centos7-py36 .`
+1. Run the build
+   * Linux: `docker run --rm -v $(pwd):/tmp/pyprt/root -w /tmp/pyprt/root pyprt:centos7-py36 python setup.py bdist_wheel`
```
[the remark this answers is not in the document]
Comment (afrischknecht) [reply README.md:154]: That said I don't know how cool it is to mount `/etc/shadow` into a black box. At least I would think twice before copy-pasting this command blindly into my shell. (If I found it in the README of a public project.)

**Review by afrischknecht — CHANGES_REQUESTED** [review]
Generally okay. The documented command did not work for me. Mapping of user ids from host to container is problematic (at least on Linux). My recommendation is to dodge that bullet entirely by

1. Removing the USER instruction from the Dockerfile.
2. Running as root inside the container (i.e. remove the `-u uid:gid` instruction)
3. Appending `; chown -R $(id -u):$(id -g) /tmp/pyprt/ws/build` to the bash -c argument (and replace single quotes with double-quotes).

i.e. the full command would be:
`docker run --rm -v $(pwd):/tmp/pyprt/ws -w /tmp/pyprt/ws/ pyprt:centos7-py36 bash -c "python setup.py bdist_wheel --dist-dir=/tmp/pyprt/ws/build; chown -R $(id -u):$(id -g) /tmp/pyprt/ws/build"`

That way there is no need to mess around with mounting passwd and shadow files into the container, yet you still end up with the whole build folder owned by the Docker host user once the container is done.

File: envs/centos7/py36/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+RUN yum install -y git cmake3 ninja-build devtoolset-8-toolchain \
+rh-python36-python rh-python36-python-devel rh-python36-python-pip
+
+RUN alternatives --install /usr/local/bin/cmake cmake /usr/bin/cmake3 20 \
+--slave /usr/local/bin/ctest ctest /usr/bin/ctest3 \
+--slave /usr/local/bin/cpack cpack /usr/bin/cpack3 \
+--slave /usr/local/bin/ccmake ccmake /usr/bin/ccmake3 \
+--family cmake
+
+USER 997:995
```
Comment (afrischknecht) [inline envs/centos7/py36/Dockerfile:14]: The naked USER directive is a bit odd without a corresponding RUN that adds a user and group with a corresponding uid/gid to the container image.

**Review by afrischknecht — CHANGES_REQUESTED** [review]

File: envs/centos7/py36/Dockerfile (hunk 2 above)
Comment (mistafunk) [reply envs/centos7/py36/Dockerfile]: agreed, see above

File: README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (mistafunk) [reply README.md:156]: i removed the need by implementing a "compile time user setup". downside is that you need to build the image per machine/user configuration. upside is no need for permission acrobatics.

inspired by https://vsupalov.com/docker-shared-permissions/

[the remark this answers is not in the document]
Comment (afrischknecht) [reply README.md:156]: That's also a good (maybe even better) approach, especially in this case.

**Review by afrischknecht — CHANGES_REQUESTED** [review]
Almost perfect. The only thing that didn't work for me was the `docker run` command in the README. I had to do
`docker run --rm -v $(pwd):/tmp/pyprt/root -w /tmp/pyprt/root pyprt:centos7-py36 bash -c 'python setup.py bdist_wheel'` (note the `bash -c '...'`) else CMake is complaining that it cannot properly detect the C++ compiler. ("No CMAKE_CXX_COMPILER could be found")

**Comment by mistafunk** [comment]
> Almost perfect. The only thing that didn't work for me was the `docker run` command in the README. I had to do
> `docker run --rm -v $(pwd):/tmp/pyprt/root -w /tmp/pyprt/root pyprt:centos7-py36 bash -c 'python setup.py bdist_wheel'` (note the `bash -c '...'`) else CMake is complaining that it cannot properly detect the C++ compiler. ("No CMAKE_CXX_COMPILER could be found")

good catch - that was actually out of sync with what the Jenkinsfile was already doing - fixed.

**Review by afrischknecht — APPROVED** [review]
Perfect. PR greenlit to merge.

**Pull request merged**